src/common-libs/plan/index.ts

```diff
@@ -0,0 +1,511 @@
+        },
+        where: { id: createBrandedValue(baseRow.id) },
+      });
+    } else {
+      inactiveRows.push(createBrandedValue(baseRow.id));
+    }
+  }
+
+  await model.update({
+    values: {
+      latestTaggedVersion: false,
+      ...(tag.public ? { currentVersion: false } : {}),
+    },
+    where: {
+      id: { [models.Op.IN]: inactiveRows.map(createBrandedValue) },
+    },
+  });
+
+  const versionRows = await versionModel.find({
+    where: {
+      [idField]: {
+        [models.Op.IN]: baseRows.map((a) => a.id),
+      },
+      createdAt: { [models.Op.LTE]: tag.createdAt },
+    },
+  });
+
+  const versionRowsByBaseId = new Map<
+    unknown,
+    Set<InstanceDataOfModel<typeof versionModel>>
+  >();
+  for (const obj of versionRows) {
+    const value = obj[idField as keyof typeof obj];
+    let group = versionRowsByBaseId.get(value);
+    if (!group) {
+      versionRowsByBaseId.set(value, (group = new Set()));
+    }
+    group.add(obj);
+  }
+
+  const latestVersionIds = [...versionRowsByBaseId.values()].map(
+    (attachmentVersions) =>
+      // Get biggest attachment version ID
+      Math.max(...[...attachmentVersions].map((av) => av.id))
+  );
+
+  const latestVersions = await versionModel.find({
+    where: {
+      id: {
+        [models.Op.IN]: latestVersionIds.map(createBrandedValue),
+      },
+    },
+  });
+
+  for (const latestVersion of latestVersions) {
+    await versionModel.update({
+      values: {
+        latestTaggedVersion: true,
+        versionTags: [
+          ...latestVersion.versionTags,
+          ...(latestVersion.versionTags.includes(tag.name) ? [] : [tag.name]),
+        ],
+        ...(tag.public ? { currentVersion: true } : {}),
+      },
+      where: { id: createBrandedValue(latestVersion.id) },
+    });
+  }
+
+  const oldVersions = await versionModel.find({
+    where: {
+      id: {
+        [models.Op.NOT_IN]: latestVersionIds.map(createBrandedValue),
+      },
+      [idField]: { [models.Op.IN]: baseRows.map((a) => a.id) },
+    },
+  });
+  await versionModel.update({
+    values: {
+      latestTaggedVersion: false,
+      ...(tag.public ? { currentVersion: false } : {}),
+    },
+    where: {
+      id: {
+        [models.Op.IN]: oldVersions.map((v) => v.id as AnyModelId),
+      },
+    },
+  });
+};
+
+const updateEntityVersionModelTags = async (
+  models: Database,
+  tableName: 'planVersion' | 'planLocation' | 'planYear',
+  tag: Pick<
+    InstanceDataOfModel<Database['planTag']>,
+    'createdAt' | 'planId' | 'public' | 'name'
+  >
+) => {
+  const model = models._tables[tableName];
+  const idField = `planId`;
+  type AnyModelId = Brand<number, any, any>;
+
+  const versionRows = await model.find({
+    where: {
+      [idField]: tag.planId,
+      createdAt: { [models.Op.LTE]: tag.createdAt },
+    },
+  });
+
+  const versionRowsByPlanId = new Map<
+    unknown,
+    Set<InstanceDataOfModel<typeof model>>
+  >();
```

Comment:
`import type { PlanId } from '@unocha/hpc-api-core/src/db/models/plan';`
```suggestion
  const versionRowsByPlanId = new Map<
    PlanId,
    Set<InstanceDataOfModel<typeof model>>
  >();
```